warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/TermFrequencyHitFunction.java

```diff
@@ -0,0 +1,689 @@
+            if (backing.byteAt(i) == '\u0000') {
+                index = i + 1;
+                break;
+            }
+        }
+        return new String(backing.subSequence(index, backing.length()).toArray());
+    }
+    
+    // parse the datatype directly from a key with a column family like datatype\0uid
+    private String parseDatatypeFromCF(Key k) {
+        ByteSequence backing = k.getColumnFamilyData();
+        int index = -1;
+        for (int i = backing.offset(); i < backing.length(); i++) {
+            if (backing.byteAt(i) == '\u0000') {
+                index = i;
+                break;
+            }
+        }
+        return new String(backing.subSequence(0, index).toArray());
+    }
+    
+    /**
+     * Preprocessing step, parse functions into their respective sub queries and precompute search space.
+     * 
+     * @param script
+     *            the query tree
+     */
+    private void populateFunctionSearchSpace(ASTJexlScript script) {
+        Multimap<String,Function> allFunctions = TermOffsetPopulator.getContentFunctions(script);
+        Set<Function> functions = new HashSet<>(allFunctions.values());
+        for (Function function : functions) {
+            List<SubQuery> subQueries = parseFunction(function);
+            for (SubQuery subQuery : subQueries) {
+                
+                // populate field search space
+                functionFields.add(subQuery.field);
+                
+                if (subQuery.isNegated) {
+                    negativeSubQueries.add(subQuery);
+                    for (String fieldValue : subQuery.fieldValues) {
+                        fvToFunctionIds.put(fieldValue, subQuery.id);
+                        negativeSearchSpace.add(fieldValue);
+                    }
+                } else {
+                    positiveSubQueries.add(subQuery);
+                    for (String fieldValue : subQuery.fieldValues) {
+                        fvToFunctionIds.put(fieldValue, subQuery.id);
+                        positiveSearchSpace.add(fieldValue);
+                    }
+                }
+            }
+        }
+    }
+    
+    /**
+     * Get this function's search space as defined by fields and values
+     *
+     * @param function
+     *            a {@link Function} that is either adjacent, phrase, or within
+     * @return a list of fields and values associated with the provided function
+     */
+    private List<SubQuery> parseFunction(Function function) {
+        // functions may take different forms..
+        // within = {field, number, termOffsetMap, terms...}
+        // within = {number, termOffsetMap, terms...}
+        // adjacent = {field, termOffsetMap, terms...}
+        // adjacent = {termOffsetMap, terms...}
+        // phrase = {field, termOffsetMap, terms...}
+        // phrase = {termOffsetMap, terms...}
+        List<JexlNode> args = function.args();
+        int index = function.name().equals("within") ? 3 : 2;
```

Comment:
It makes me feel a little strange that we go to the trouble to enumerate our expectations here in a comment, but we don't entirely enforce them in code. I would certainly feel more comfortable if we enforced that the function was one of 'within', 'adjacent' or 'phrase' if those are truly the only functions we plan to handle, and then somehow gracefully degrade if we don't understand the function we've encountered.

Reply:
This is a reasonable change, I'll make it so.